In [2]:
# binomial single 
# prob of event
# binomial summation (tails)

In [3]:
# explain what discrete means

In [8]:
# basic functions
import math
from scipy.special import comb
from scipy.stats import binom

In [9]:
# create visuals and visual style
import matplotlib.pyplot as plt
import seaborn as sns

sns.set(style='white', context='notebook', palette='deep')

%matplotlib inline

**Scipy** has a module called [scipy.stats](https://docs.scipy.org/doc/scipy/reference/stats.html) that has a binomial distribution function under the header **Discrete Distributions**

[Scipy.special](https://docs.scipy.org/doc/scipy/reference/special.html)

[Scipy.stats Docs](https://docs.scipy.org/doc/scipy/reference/stats.html) 

In [5]:
numbers = [0,1,2,3,4,5,6,7]

for item in numbers:
    print(f"{item}! = {math.factorial(item):,}")

0! = 1
1! = 1
2! = 2
3! = 6
4! = 24
5! = 120
6! = 720
7! = 5,040


On **page 37** of this chapter, we're introduced to **combinatorics** or *"advanced counting"*.

This is commonly presented in the form ***INSERT LATEX***

Which reads "*n choose k*". What this expresses is the number of k combinations in n items.In R, this is the choose() function.

**Scipy.special** has a similar method called **comb** that does the same thing.

In [6]:
comb(N=3,k=2,exact=True) # exact=True returns an integer, exact=False float

3

**Scipy.stats** uses [**binom.pmf**](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binom.html#scipy.stats.binom) to calculate the binomial *probability mass function*.

On **page 39** is the example:

> *We could calculate the probability of flipping exactly 12 heads in 24 coin tosses.*

binom.pmf(k,n,p)

k = 12
n = 24
p = 0.5

In [7]:
probability = binom.pmf(12,24,0.5)

round(probability,4) # round returns an integer to four decimal places

0.1612

In [11]:
nums = list(range(0,24+1))
probs = [binom.pmf(n,24,0.5) for n in nums]

### Example: Gacha Games

> Is the probability of getting a E.T. Jaynes card greater than or equal to 0.5?

**ADD LATEX HERE**

number of items (n) = 100

number of ET Jaynes cards (k) is *greater than or equal to one*

random chance of getting ET Jaynes card (p) = 0.720% (not 72.0%!)

There are two ways to approach this problem using scipy:
1. Use binom.sf "survival function"
2. Use binom.cdf "cumulative density function"

We'll use binom.sf first because it is the most similar to the example illustrated in the book. 

The pmf calculates the probability for a specific number k. But since we are only concerned with getinng *one ET Jaynes card **or more*** we want to calculate a cumulative probability. So whether we get 1 card or 20 is OK. So essentially, what we are doing is calculating the pmf for every number up to 100 and adding all of those together.

survival function prob at the random variable is larger (but not) equal to k

binom.sf arguments:

k = this is one less than minimum card we want (1) so 0.

n = number of items (n) or 100

p = 0.720% or 0.00720

what does loc argument mean?

In [48]:
result = binom.sf(0,100,0.00720)

f'{result:.5f} >= 0.5'

'0.51451 >= 0.5'

Since 51.4% is greater than 50.0%, buy the cards

The other way to go about this is to use binom.cdf

What binom.cdf essentially says is we are looking for *the cumulative density up to a number*'. In this case that number is zero because if we get one, we're happy.

The difference is that in this method, we're calculating the amount below our number (1) which is that chance we won't get one card.

So we have to subtract this from 1 to get the 0.51451 answer.

In [49]:
1- binom.cdf(1,100,0.0072)

0.16242880152251793

Bonus example: What are the chances of getting at least two ET Jaynes cards?

In [50]:
sf_chance = binom.sf(1,100,0.0072)
cdf_chance = 1 - binom.cdf(1,100,0.0072)
print(f'sf approach: {sf_chance:.5f}')
print(f'cdf approach: {cdf_chance:.5f}')

sf approach: 0.16243
cdf approach: 0.16243


### Summary of Python used:

combinations: comb(N,k,exact=True)

binomial probability mass function: binom.pmf(k,n,p)

binomial survival function: binom.sf(k,n,p)

binomial cumulative density function: binom.cdf(k,n,p)

### Questions

In [93]:
def binom_probs(n,p):
    seq = [n for n in range(0,n+1)]
    return [binom.pmf(s,n,p) for s in seq]

In [95]:
tt = binom_probs(7,0.2);
tt

[0.20971519999999993,
 0.36700160000000004,
 0.27525120000000014,
 0.11468800000000005,
 0.028672000000000024,
 0.004300800000000003,
 0.0003584000000000004,
 1.28e-05]

In [85]:
binom.cdf(3,6,0.5, loc=0)

0.6562499999999999

In [ ]:
# 3 is our k
# 6 is number of tries
# 0.5 is our p
# loc is where in the array we start from...verify with docs

what this says is the chance of "at least two offers" only scenarios are
less than 2 (ie 0 or 1 offers). so binom cdf calculates the array of pdf probs for all seven attempts. then it calculates 

In [ ]:
Lesson idea: How to find a probability problem.